src/group/managed_group/mod.rs

```diff
@@ -0,0 +1,276 @@
+/// This allows the Delivery Service to reject them (e.g. if they reference the wrong epoch).
+///
+/// If incoming messages or applied operations are semantically or syntactily incorrect, the function will return a corresponding
+/// error message and the state of the group will remain unchanged.
+///
+/// The application policy for the group can be enforced by implementing the validator callback functions and selectively allowing/
+/// disallowing each operation (see `ManagedGroupCallbacks`)
+///
+/// Changes to the group state are dispatched as events through callback functions (see ManagedGroupCallbacks).
+pub struct ManagedGroup {
+    managed_group_config: ManagedGroupConfig,
+    group: MlsGroup,
+    proposal_queue: ProposalQueue,
+    own_kpbs: Vec<KeyPackageBundle>,
+}
+
+impl ManagedGroup {
+    // === Group creation ===
+
+    /// Creates a new group from scratch with only the creator as a member.
+    pub fn new(
+        credential_bundle: &CredentialBundle,
+        managed_group_config: &ManagedGroupConfig,
+        group_id: GroupId,
+        ciphersuite_name: CiphersuiteName,
+        key_package_bundle: KeyPackageBundle,
+    ) -> Result<Self, GroupError> {
+        let group = MlsGroup::new(
+            &group_id.as_slice(),
+            ciphersuite_name,
+            key_package_bundle,
+            GroupConfig::default(),
+        )?;
+
+        Ok(ManagedGroup {
+            managed_group_config: managed_group_config.clone(),
+            group,
+            proposal_queue: ProposalQueue::new(),
+            own_kpbs: vec![],
+        })
+    }
+
+    /// Creates a new group from a `Welcome` message
+    pub fn new_from_welcome(
+        credential_bundle: &CredentialBundle,
+        managed_group_config: &ManagedGroupConfig,
+        welcome: Welcome,
+        ratchet_tree: Option<Vec<Option<Node>>>,
+        key_package_bundle: KeyPackageBundle,
+    ) -> Result<Self, WelcomeError> {
+        let group = MlsGroup::new_from_welcome(welcome, ratchet_tree, key_package_bundle)?;
+        Ok(ManagedGroup {
+            managed_group_config: managed_group_config.clone(),
+            group,
+            proposal_queue: ProposalQueue::new(),
+            own_kpbs: vec![],
+        })
+    }
+
+    /// Creates a new group with the creator and additional members
+    pub fn new_with_members(
+        credential_bundle: &CredentialBundle,
+        managed_group_config: &ManagedGroupConfig,
+        group_id: GroupId,
+        ciphersuite_name: CiphersuiteName,
+        key_package_bundle: KeyPackageBundle,
+        members: Vec<KeyPackage>,
+    ) -> Result<Self, GroupError> {
+        unimplemented!()
+    }
+
+    // === Membership management ===
+
+    /// Adds members to the group
+    pub fn add_members(
+        &mut self,
+        credential_bundle: &CredentialBundle,
+        aad: &[u8],
+        key_packages: Vec<KeyPackage>,
```

Comment:
We should add some sort of client/member struct that we can use here to consistently identify clients/members. Then we can use the same type for add, remove, etc.

Reply:
It's tricky, because the amount of information varies:
 - Adds: You need a full KeyPackage
 - Remove: You just need an index
 - Uniquely referencing a member: You need a Credential

We could refactor remove to accept a credential instead of an index, but the problem with Adds remains.

Reply:
I created #213